ksql-engine/src/main/java/io/confluent/ksql/analyzer/QueryAnalyzer.java

```diff
@@ -130,6 +133,23 @@ public AggregateAnalysis analyzeAggregate(final Query query, final Analysis anal
     return aggregateAnalysis;
   }
 
+  public TableFunctionAnalysis analyzeTableFunctions(final Analysis analysis) {
```

Comment:
I am open to that, the only reason I did it this way, is that's the way it was does for aggregates so I guessed that was the preferred way. I suggest this is done as a further refactoring so as not to make this one more complex.

Reply:
Just to clarify, are you suggesting that we get rid of TableFunctionAnalyzer and AggregateAnalyzer and only have Analyzer?

Currently, Analyzer only pulls out select expressions, group by expressions, having expressions and the like. It does not look inside the actual expressions to find the actual aggregate or table function calls.

Currently it's table function analyzer and aggregate analyzer that actually look inside the expressions to find the actual functions.

I think it might be better if Analyzer remains "dumb" and simply extracts the high level information from the query without knowing about the specifics of aggregation and table function. It seems that the knowledge of what table functions are, and of what aggregate functions are belongs in FlatMapNode and AggregateNode for better separation of concerns.

So how about we get rid of both table function analyzer and aggregate analyzer and move the extraction of table functions and aggregate functions, and rewriting of select expressions into FlatMapNode and AggregateNode? It seems right now we have this weird split where some of the logic of computing flat mapping is in table function analyzer and the rest is in FlatMapNode. Imho it should be encapsulated in one place - FlatMapNode. Similar reasoning applies for aggregation. I'd recommend keeping Analyzer dumb - just extracting the information from the query without knowing what the information is for.

Reply:
Hummm... I think we need to be careful of moving logic into the Logical model, e.g.  `FlatMapNode` and `AggregateNode`.

IMHO, the logical model should be dumb pojos.  The smarts should be in things that build / manipulate or process the model.

That's not to say I think the current mode of operation is ideal! But that probably requires more thought, outside of this PR / feature.  

To be clear, I was asking above if it was possible to capture details of UDTfs in the same place we capture info on UDFs.  I don't think we need another analyser just for UDTFs.

Reply:
https://github.com/confluentinc/ksql/issues/3673 As discussed on zoom